django-stubs/contrib/admin/decorators.pyi

```diff
@@ -1,27 +1,50 @@
 from collections.abc import Callable, Sequence
-from typing import Any, TypeVar
+from typing import Any, TypeVar, overload
 
 from django.contrib.admin import ModelAdmin
 from django.contrib.admin.sites import AdminSite
 from django.db.models import Combinable, QuerySet
 from django.db.models.base import Model
 from django.db.models.expressions import BaseExpression
 from django.http import HttpRequest
+from django.utils.functional import _StrPromise
 
-_ModelT = TypeVar("_ModelT", bound=Model)
+_T = TypeVar("_T")
+_Model = TypeVar("_Model", bound=Model)
+_ModelAdmin = TypeVar("_ModelAdmin", bound=ModelAdmin)
+_Request = TypeVar("_Request", bound=HttpRequest)
+_QuerySet = TypeVar("_QuerySet", bound=QuerySet)
 
+@overload
+def action(
+    function: Callable[[_ModelAdmin, _Request, _QuerySet], None],
+    permissions: Sequence[str] | None = ...,
+    description: _StrPromise | None = ...,
+) -> Callable[[_ModelAdmin, _Request, _QuerySet], None]: ...
```

Comment:
Would be easier and less repetition to define

```python
_ActionCallable = TypeVar("_ActionCallable", bound=Callable[[ModelAdmin, HttpRequest, QuerySet], None])
```

Reply:
That's not gonna work. It needs a bound, if you declare them without a `TypeVar` it's gonna change the signature of the function that is decorated

Reply:
Oh, right, the bound callable won't match either. I started out with that but then you'll need nested TypeVars to get a correct bound. Else we're not supporting subclasses. See here

https://mypy-play.net/?mypy=0.982&python=3.10&gist=de1239d0a4ced1707ec6afb9f4610160

Reply:
It looks cluttered, I agree, but this was the only way I could get it to work correctly

Reply:
Your example causes an error because you have `B` and `A` flipped.

`def decorated(obj: A) -> str` is compatible with `Callable[[B], Any]`, but your example has these the other way around.

Here's a working example. I removed the `_A` TypeVar too: https://mypy-play.net/?mypy=0.990&python=3.11&gist=a5f4a634dba1b01c20cf831662de471f

Reply:
But `A` represents `models.Model`, `decorated` is my custom display function and `B` is one of the models from my project.

So, in your playground link, it's only possible to have `obj: models.Model` which would give me an error if I try to access a field on one of my models

Reply:
https://mypy-play.net/?mypy=0.990&python=3.11&gist=5670bdd2e498f5159d47e26a41d653d5

Reply:
> [...] Which is the correct subclass in the particular instance won't be type-checked.
>
> It's a bit of a hack and not entirely type-safe, but I think practicality beats purity.

Could you elaborate on what it is that won't be type checked and why it's not type safe? (I presume with the current display annotations `func: Callable[[_ModelT], _T]`)

I'm not sure I'm following on how that conclusion is reached.

Reply:
For example given two models `Foo` and `Bar`. With your proposed solution, you may have a `FooAdmin` class with a `@display` method `def my_display_func(obj: Bar)`. The type checker has no idea that `Bar` is incorrect and the method will actually accept `Foo` instances. That's what I meant by "not type-safe"

The "type-safe" approach is contravariant compatibility: you can define `my_display_func(obj: Model)` or `my_display_func(obj: object)`, and it's assured that any passed argument is always an instance of `Model` or `object`. It's type-safe, but useless: almost every display callback needs to access fields of the particular Model subclass, you'd need to use `cast()` or `isinstance()` guards to do most useful things.

Reply:
Hm yeah that's true. Didn't think of that. 

I suppose in order to get around that, it would be necessary for `admin.display` to have the context of `FooAdmin` (e.g. `FooAdmin.model`), although it's a naive, dynamic, decorator that doesn't even consider the ModelAdmin.

What I'm saying is e.g. `@admin.display[MyModelAdmin]` (if ModelAdmin accepted arguments runtime) and pick up its model. But then it also has to be possible to pick up the Model argument for ModelAdmin